sklearn/utils/estimator_checks.py

```diff
@@ -2345,3 +2347,60 @@ def check_outliers_fit_predict(name, estimator_orig):
         for contamination in [-0.5, 2.3]:
             estimator.set_params(contamination=contamination)
             assert_raises(ValueError, estimator.fit_predict, X)
+
+
+def check_fit_idempotent(name, estimator_orig):
+    # Check that est.fit(X) is the same as est.fit(X).fit(X). Ideally we would
+    # check that the estimated parameters during training (e.g. coefs_) are
+    # the same, but having a universal comparison function for those
+    # attributes is difficult and full of edge cases. So instead we check that
+    # predict(), predict_proba(), decision_function() and transform() return
+    # the same results.
+
+    est = clone(estimator_orig)
+
+    np.random.seed(0)
+    X = np.random.normal(loc=100, size=(100, 2))
+    if is_regressor(estimator_orig):
+        y = np.random.normal(size=100)
+    else:
+        y = np.random.randint(low=0, high=2, size=100)
+    if est.__class__.__name__.startswith('MultiTask'):
+        y = np.stack([y, y], axis=1)
+
+    X_train, X_test, y_train, _ = train_test_split(X, y)
+
+    if 'random_state' in est.get_params().keys():
+        est.set_params(random_state=0)
+    if 'warm_start' in est.get_params().keys():
+        est.set_params(warm_start=False)
+
+    est.fit(X_train, y_train)
+
+    if hasattr(est, 'predict'):
+        pred_1 = est.predict(X_test)
+    if hasattr(est, 'predict_proba'):
+        pred_proba_1 = est.predict_proba(X_test)
+    if hasattr(est, 'decision_function'):
+        decision_1 = est.decision_function(X_test)
+    if hasattr(est, 'transform'):
+        transform_1 = est.transform(X_test)
+
+    # Fit again
+    est.fit(X_train, y_train)
+
+    if hasattr(est, 'predict'):
```

Comment:
Can we consider resuscitating the concept of assert_same_model (#4841)?

Reply:
Ideally yes...

I tried to do that by comparing the attributes ending with `_` but ran into tons of edge cases, so I went with the prediction comparison, which is probably enough for this issue.

Basically the question is, is it really worth it? It's doable but there's no doubt the code is going to be a pain to maintain.

The code in #4841 ignores the comparison of some attributes, in particular the attributes that are themselves estimators (see [here](https://github.com/scikit-learn/scikit-learn/pull/4841/files#diff-f0af7b240082b7b0e03078a38121ae1aR641)), and it's already pretty involved!

Reply:
I think most of the time we can presume that identical predictions/transformation is strong enough (although if all attributes can be tested for equality that would be faster in many cases). Where those methods do not exist, then we may need an alternative.
Asserting that two models are not equal would be harder. A function is_model_equal returning {yes, likely, no, unknown} would be useful and could be strengthened over time

Reply:
So what do you suggest exactly? Should I turn this test into `is_model_equal` and return either likely if it passes or unknown if it fails?

Reply:
If transform etc are not equal we know the model is not equal. If we can test that the objects are identical we know the models are identical. If we show that predict, transform, etc, match on random data then it is likely they are equal, and if we have no way to test, then it is unknown.

Here the test would pass if anything but 'no' was returned.

Don't we have similar assertions elsewhere? I'd expect this to be used multiple times already.

Reply:
>Don't we have similar assertions elsewhere?

I have no idea to be honest.

I still don't understand what you want to do regarding this PR:

- Do you want to have `is_model_equal`?
- If yes, in this PR? Another one? Should I create a new issue so we can discuss requirements / implementation details there?
- Should we still merge this PR, and if no do you want to reimplement it once we have `is_model_equal`?

Reply:
Omg. I'm sorry. I was coming at this thinking that we had tests with similar checks of model equivalence... But maybe they were only in pull requests, or my imagination, or aren't expressed like this. Okay. Let's not bother about making it generic in this issue.